### PR TITLE
Trigger GC loop when queue is purged or deleted

### PR DESCRIPTION
**Pull request by snichme**

### CODE REVIEW

**Comment by snichme** [comment]
Fixes #201

**Review by antondalgren — APPROVED** [review]

File: spec/server_spec.cr [hunk 1]
```diff
@@ -141,15 +141,6 @@ describe AvalancheMQ::Server do
     end
   end
 
-  it "should mark the vhost as dirty when a queue is purged" do
```
Comment (carlhoerberg) [inline spec/server_spec.cr:144]: Can we write another test for this? but one that checks that bytes are actually recovered from disk or something?

File: src/avalanchemq/queue/queue.cr [hunk 2]
```diff
@@ -825,6 +826,7 @@ module AvalancheMQ
       count = @ready.purge
       @log.debug { "Purged #{count} messages" }
       @vhost.dirty = true
+      @vhost.gc_loop.send nil # trigger gc for message clean up
```
Comment (carlhoerberg) [inline src/avalanchemq/queue/queue.cr:829]: how about a new method called `trigger_gc` that sets the dirty flag and sends nil to the channel?

File: src/avalanchemq/vhost.cr [hunk 3]
```diff
@@ -672,32 +673,39 @@ module AvalancheMQ
...
-        unless @dirty
-          sleep Config.instance.gc_segments_interval
-          next
+        select
+        when @gc_loop.receive
+          run_gc(referenced_sps)
+        when timeout interval
+          next unless @dirty
+          run_gc(referenced_sps)
         end
```
Comment (carlhoerberg) [inline src/avalanchemq/vhost.cr:686]:
```suggestion
        select
        when @gc_loop.receive
        when timeout interval
        end
        run_gc(referenced_sps)
```

File: src/avalanchemq/vhost.cr [hunk 4]
```diff
@@ -672,32 +673,39 @@ module AvalancheMQ
...
+                   "killing process #{ex.inspect_with_backtrace}")
+        exit 1
       end
-    rescue ex
-      @log.fatal("Unhandled exception in #gc_segments_loop, killing process #{ex.inspect_with_backtrace}")
-      exit 1
+    end
+
+    private def run_gc(referenced_sps)
+      return if @closed
```
Comment (carlhoerberg) [inline src/avalanchemq/vhost.cr:695]:
```suggestion
      return if @closed
      return unless @dirty
```

File: spec/server_spec.cr (hunk 1 above)
Comment (snichme) [reply spec/server_spec.cr:144]: Yes, some type of test that verifies that GC has run. Iäll take a look at it.

**Pull request merged**